From bcc91a88b06ae70cd39f84e616d07e1f624c54d4 Mon Sep 17 00:00:00 2001
From: benrichard-amd <ben.richard@amd.com>
Date: Tue, 25 Aug 2026 16:32:56 +0000
Subject: [PATCH] gallium/ddebug: exclude ddebug from non-gfx builds

Move functions declared in dd_util.h from dd_draw.c into a new
dd_util.c. This allows dd_draw.c to move entirely into the
with_gfx_compute block alongside dd_context.c and dd_screen.c.

Fixes a link failure on ppc64le when building with decode-only VA.
---
 src/gallium/auxiliary/driver_ddebug/dd_draw.c |  77 ------------
 src/gallium/auxiliary/driver_ddebug/dd_util.c | 116 ++++++++++++++++++
 src/gallium/auxiliary/driver_ddebug/dd_util.h |   1 +
 src/gallium/auxiliary/meson.build             |   3 +-
 4 files changed, 119 insertions(+), 78 deletions(-)
 create mode 100644 src/gallium/auxiliary/driver_ddebug/dd_util.c

diff --git a/src/gallium/auxiliary/driver_ddebug/dd_draw.c b/src/gallium/auxiliary/driver_ddebug/dd_draw.c
index 36bf5b03da0..2709fb69402 100644
--- a/src/gallium/auxiliary/driver_ddebug/dd_draw.c
+++ b/src/gallium/auxiliary/driver_ddebug/dd_draw.c
@@ -42,83 +42,6 @@
 #include "util/detect.h"
 #include "c99_alloca.h"
 
-void
-dd_get_debug_filename_and_mkdir(char *buf, size_t buflen, bool verbose)
-{
-   static unsigned index;
-   char dir[256];
-   const char *proc_name = util_get_process_name();
-
-   if (!proc_name) {
-      fprintf(stderr, "dd: can't get the process name\n");
-      proc_name = "unknown";
-   }
-
-   snprintf(dir, sizeof(dir), "%s/"DD_DIR, debug_get_option("HOME", "."));
-
-   if (os_mkdir(dir, 0774) && errno != EEXIST)
-      fprintf(stderr, "dd: can't create a directory (%i)\n", errno);
-
-   snprintf(buf, buflen, "%s/%s_%u_%08u", dir, proc_name, (unsigned int)getpid(),
-            (unsigned int)p_atomic_inc_return(&index) - 1);
-
-   if (verbose)
-      fprintf(stderr, "dd: dumping to file %s\n", buf);
-}
-
-FILE *
-dd_get_debug_file(bool verbose)
-{
-   char name[512];
-   FILE *f;
-
-   dd_get_debug_filename_and_mkdir(name, sizeof(name), verbose);
-   f = fopen(name, "w");
-   if (!f) {
-      fprintf(stderr, "dd: can't open file %s\n", name);
-      return NULL;
-   }
-
-   return f;
-}
-
-void
-dd_parse_apitrace_marker(const char *string, int len, unsigned *call_number)
-{
-   unsigned num;
-   char *s;
-
-   if (len <= 0)
-      return;
-
-   /* Make it zero-terminated. */
-   s = alloca(len + 1);
-   memcpy(s, string, len);
-   s[len] = 0;
-
-   /* Parse the number. */
-   errno = 0;
-   num = strtol(s, NULL, 10);
-   if (errno)
-      return;
-
-   *call_number = num;
-}
-
-void
-dd_write_header(FILE *f, struct pipe_screen *screen, unsigned apitrace_call_number)
-{
-   char cmd_line[4096];
-   if (util_get_command_line(cmd_line, sizeof(cmd_line)))
-      fprintf(f, "Command: %s\n", cmd_line);
-   fprintf(f, "Driver vendor: %s\n", screen->get_vendor(screen));
-   fprintf(f, "Device vendor: %s\n", screen->get_device_vendor(screen));
-   fprintf(f, "Device name: %s\n\n", screen->get_name(screen));
-
-   if (apitrace_call_number)
-      fprintf(f, "Last apitrace call: %u\n\n", apitrace_call_number);
-}
-
 FILE *
 dd_get_file_stream(struct dd_screen *dscreen, unsigned apitrace_call_number)
 {
diff --git a/src/gallium/auxiliary/driver_ddebug/dd_util.c b/src/gallium/auxiliary/driver_ddebug/dd_util.c
new file mode 100644
index 00000000000..a074aba466b
--- /dev/null
+++ b/src/gallium/auxiliary/driver_ddebug/dd_util.c
@@ -0,0 +1,116 @@
+/**************************************************************************
+ *
+ * Copyright 2015 Advanced Micro Devices, Inc.
+ * Copyright 2008 VMware, Inc.
+ * All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * on the rights to use, copy, modify, merge, publish, distribute, sub
+ * license, and/or sell copies of the Software, and to permit persons to whom
+ * the Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHOR(S) AND/OR THEIR SUPPLIERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+ * USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ **************************************************************************/
+
+#include "dd_util.h"
+
+#include "pipe/p_screen.h"
+#include "util/os_file.h"
+#include "util/u_atomic.h"
+#include "util/u_process.h"
+#include "c99_alloca.h"
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+void
+dd_get_debug_filename_and_mkdir(char *buf, size_t buflen, bool verbose)
+{
+   static unsigned index;
+   char dir[256];
+   const char *proc_name = util_get_process_name();
+
+   if (!proc_name) {
+      fprintf(stderr, "dd: can't get the process name\n");
+      proc_name = "unknown";
+   }
+
+   snprintf(dir, sizeof(dir), "%s/"DD_DIR, debug_get_option("HOME", "."));
+
+   if (os_mkdir(dir, 0774) && errno != EEXIST)
+      fprintf(stderr, "dd: can't create a directory (%i)\n", errno);
+
+   snprintf(buf, buflen, "%s/%s_%u_%08u", dir, proc_name, (unsigned int)getpid(),
+            (unsigned int)p_atomic_inc_return(&index) - 1);
+
+   if (verbose)
+      fprintf(stderr, "dd: dumping to file %s\n", buf);
+}
+
+FILE *
+dd_get_debug_file(bool verbose)
+{
+   char name[512];
+   FILE *f;
+
+   dd_get_debug_filename_and_mkdir(name, sizeof(name), verbose);
+   f = fopen(name, "w");
+   if (!f) {
+      fprintf(stderr, "dd: can't open file %s\n", name);
+      return NULL;
+   }
+
+   return f;
+}
+
+void
+dd_parse_apitrace_marker(const char *string, int len, unsigned *call_number)
+{
+   unsigned num;
+   char *s;
+
+   if (len <= 0)
+      return;
+
+   /* Make it zero-terminated. */
+   s = alloca(len + 1);
+   memcpy(s, string, len);
+   s[len] = 0;
+
+   /* Parse the number. */
+   errno = 0;
+   num = strtol(s, NULL, 10);
+   if (errno)
+      return;
+
+   *call_number = num;
+}
+
+void
+dd_write_header(FILE *f, struct pipe_screen *screen, unsigned apitrace_call_number)
+{
+   char cmd_line[4096];
+   if (util_get_command_line(cmd_line, sizeof(cmd_line)))
+      fprintf(f, "Command: %s\n", cmd_line);
+   fprintf(f, "Driver vendor: %s\n", screen->get_vendor(screen));
+   fprintf(f, "Device vendor: %s\n", screen->get_device_vendor(screen));
+   fprintf(f, "Device name: %s\n\n", screen->get_name(screen));
+
+   if (apitrace_call_number)
+      fprintf(f, "Last apitrace call: %u\n\n", apitrace_call_number);
+}
diff --git a/src/gallium/auxiliary/driver_ddebug/dd_util.h b/src/gallium/auxiliary/driver_ddebug/dd_util.h
index d7b98cbf442..82e84028b07 100644
--- a/src/gallium/auxiliary/driver_ddebug/dd_util.h
+++ b/src/gallium/auxiliary/driver_ddebug/dd_util.h
@@ -28,6 +28,7 @@
 #ifndef DD_UTIL_H
 #define DD_UTIL_H
 
+#include <stdbool.h>
 #include <stdio.h>
 
 struct pipe_screen;
diff --git a/src/gallium/auxiliary/meson.build b/src/gallium/auxiliary/meson.build
index e4815f8ae07..0565598f444 100644
--- a/src/gallium/auxiliary/meson.build
+++ b/src/gallium/auxiliary/meson.build
@@ -2,7 +2,7 @@
 # SPDX-License-Identifier: MIT
 
 files_libgallium = files(
-  'driver_ddebug/dd_draw.c',
+  'driver_ddebug/dd_util.c',
   'driver_ddebug/dd_util.h',
   'pipebuffer/pb_buffer_fenced.c',
   'pipebuffer/pb_buffer_fenced.h',
@@ -178,6 +178,7 @@ if with_gfx_compute
     'draw/draw_vs.h',
     'draw/draw_vs_variant.c',
     'driver_ddebug/dd_context.c',
+    'driver_ddebug/dd_draw.c',
     'driver_ddebug/dd_pipe.h',
     'driver_ddebug/dd_public.h',
     'driver_ddebug/dd_screen.c',
-- 
2.52.0

